Clear 'cc' in nvim_open_win 'minimal' style #11361

bluz71 opened the pull request:
Add 'colorcolumn' to the list of options that should be cleared when creating
a 'minimal'-style floating window.

bluz71 committed:
Clear 'cc' in nvim_open_win 'minimal' style #11361

Add 'colorcolumn' to the list of options that should be cleared when creating
a 'minimal'-style floating window.

justinmk commented:
thanks. Lots of tests appear to need updating.

bluz71 commented:
It looks like my change broke some tests.

I will need help, since all of this is unknown territory to myself.

KillTheMule commented:
> I will need help, since all of this is unknown territory to myself.

Ok, here's something to get you started. If you need more help, the gitter channel is a nice way to contact others.

Here's the first failed test from travis:

```
[FAILED  ] test/functional/ui/float_spec.lua @ 398: floating windows with ext_multigrid can use 'minimal' style

test/functional/ui/float_spec.lua:408: Expected screen height 22 differs from actual height 17.

Expected:

  |## grid 1

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [3:----------------------------------------]|

  |## grid 2

  |*  {19:   }{20:  1 }{21:^x                                }|

  |*  {19:   }{14:  2 }y                                |

  |*  {19:   }{14:  3 }                                 |

  |  {0:~                                       }|

  |  {0:~                                       }|

  |  {0:~                                       }|

  |## grid 3

  |                                          |

  |*## grid 4

  |*  {15:x                   }|

  |*  {15:y                   }|

  |*  {15:                    }|

  |*  {15:                    }|

Actual:

  |## grid 1

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [2:----------------------------------------]|

  |  [3:----------------------------------------]|

  |## grid 2

  |*  {19:   }{20:  1 }{UNEXPECTED background = Screen.colors.LightRed:^x}{21:                                }|

  |*  {19:   }{14:  2 }{UNEXPECTED background = Screen.colors.LightRed:y}                                |

  |*  {19:   }{14:  3 }{UNEXPECTED background = Screen.colors.LightRed: }                                |

  |  {0:~                                       }|

  |  {0:~                                       }|

  |  {0:~                                       }|

  |## grid 3

  |                                          |

To print the expect() call that would assert the current screen state, use

screen:snapshot_util(). In case of non-deterministic failures, use

screen:redraw_debug() to show all intermediate screen states.  

stack traceback:

	test/functional/ui/screen.lua:579: in function '_wait'

	test/functional/ui/screen.lua:361: in function 'expect'

	test/functional/ui/float_spec.lua:408: in function <test/functional/ui/float_spec.lua:398>
```

See what the screen was expected to look like, and how it actually looked (something like `{2: ...}` says the `...` where highlighted by color 2, as defined somewhere in `test/functional/ui/float_spec.lua`). The test itself is [here](https://github.com/neovim/neovim/blob/master/test/functional/ui/float_spec.lua#L398), you can read what it does, and determine if the actual output is really what should be there (note: In this case, it's 5 lines less than what was expected before, it says so above). If not, you've found yourself a bug and need to solve that (most of the time, it's easy to reproduce it from the test, so one can work on it). If it's ok, replace the `screen:expect(...)` call [here](https://github.com/neovim/neovim/blob/master/test/functional/ui/float_spec.lua#L407) by just `screen:snapshot_util()` and run the test again, it will give you the needed `screen:expect` call to c&p it into the test to make it pas.

bfredl commented on src/nvim/window.c:
```diff
@@ -633,6 +633,16 @@ void win_set_minimal_style(win_T *wp)
+    xfree(wp->w_p_cc);
+    wp->w_p_cc = NULL;
+  }
+  if (wp->w_p_cc_cols) {
```
This is not needed. Window options are always interpreted after this function is called.

bluz71 replied:
Ok, I assume that to mean it’s a noop?
So is it not possible to do what I want to do which is to clear out cc for the minimal style?
Or do I shift the logic somewhere else?

bfredl replied:
I did not say anything about it being impossible. You clear out `wp->w_p_cc` above, that is all that is needed. We already have code that translates from `w_p_cc` to `w_p_cc_cols`, this need not be copied and pasted to every specific place `w_p_cc` is being changed. This is just like the other string options this function already changes.

bluz71 committed:
Refine clear 'cc' in nvim_open_win 'minimal' style

- No need to clear w_p_cc_cols, it will be auto-derived from w_p_cc

- Set the value of w_p_cc to "", not NULL

- Don't do any 'cc' clearing if the value of 'cc' is already empty

bluz71 commented:
Damn, test is still failing.

```viml
Actual:
  |## grid 1
  |  [2:----------------------------------------]|
  |  [2:----------------------------------------]|
  |  [2:----------------------------------------]|
  |  [2:----------------------------------------]|
  |  [2:----------------------------------------]|
  |  [2:----------------------------------------]|
  |  [3:----------------------------------------]|
  |## grid 2
  |*  {19:   }{20:  1 }{UNEXPECTED background = Screen.colors.LightRed:^x}{21:                                }|
  |*  {19:   }{14:  2 }{UNEXPECTED background = Screen.colors.LightRed:y}                                |
  |*  {19:   }{14:  3 }{UNEXPECTED background = Screen.colors.LightRed: }                                |
  |  {0:~                                       }|
  |  {0:~                                       }|
  |  {0:~                                       }|
  |## grid 3
  |                                          |
  |## grid 4
  |  {15:x                   }|
  |  {15:y                   }|
  |  {15:                    }|
  |  {15:                    }|
```

Any ideas why?
I am way out of my depth when we are dealing with the multigrid and why `grid 2` is complaining about unexpected background color.

The latest commit did fix the segfault issue, and with my simple floating window work it does clear the `cc`.

Guidance sought. Thanks.

bfredl commented:
Failure is expected, because you changed the setup of the test. As killthemule said you should place a `screen:snapshot_util()` before the expect that failed. It will output the new `expect()` code block that you should replace the expect with. 

It might include a large `default_attrs` block in the output, then copy the new attribute to the `local attrs = { ... }` block in the beginning of the file, and rerun `screen:snapshot_util()`

bluz71 committed:
Minimal floating window tests now pass

Extended the attrs and fixed the Grid failures.

bluz71 committed:
Remove duplicate attr to see if it fixes TravisCI

bluz71 commented:
Looks like my changes are finally passing on TravisCI.

`build.sr.ht` failed, but not related to this PR.

bfredl commented:
Nice, merged!

bluz71 commented:
Cool, thanks for the assist.
